src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs

```diff
@@ -176,6 +176,107 @@ protected void LookupMembersInternal(LookupResult result, NamespaceOrTypeSymbol
             }
         }
 
+#nullable enable
+        protected void LookupImplicitExtensionMembersInSingleBinder(LookupResult result, TypeSymbol type,
+            string name, int arity, ConsList<TypeSymbol>? basesBeingResolved, LookupOptions options,
+            Binder originalBinder, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+        {
+            Debug.Assert(!type.IsTypeParameter());
+            Debug.Assert(name is not null);
+
+            var compatibleExtensions = ArrayBuilder<NamedTypeSymbol>.GetInstance();
+
+            GetCompatibleExtensions(this, type, compatibleExtensions, originalBinder, basesBeingResolved, ref useSiteInfo);
+
+            var tempResult = LookupResult.GetInstance();
+            foreach (NamedTypeSymbol extension in compatibleExtensions)
+            {
+                // No need for "diagnose" since we discard the lookup results (and associated diagnostic info)
+                // unless the results are good.
+                LookupMembersInClass(tempResult, extension, name, arity, basesBeingResolved,
+                    options, originalBinder, diagnose: false, ref useSiteInfo);
+
+                // PROTOTYPE we should prefer extension members that apply to a more specific type
+                result.MergeEqual(tempResult);
+                tempResult.Clear();
+            }
+
+            tempResult.Free();
+            compatibleExtensions.Free();
+        }
+
+        // Note: we return the compatible extensions in order:
+        // for any given pair, if one is more specific it will precede the other in the list
+        private static void GetCompatibleExtensions(Binder binder, TypeSymbol type, ArrayBuilder<NamedTypeSymbol> compatibleExtensions,
+            Binder originalBinder, ConsList<TypeSymbol>? basesBeingResolved, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+        {
+            if (type.IsErrorType())
+            {
+                return;
+            }
+
+            var extensions = ArrayBuilder<NamedTypeSymbol>.GetInstance();
+            binder.GetExtensionContainers(extensions, originalBinder);
+
+#if !DEBUG
+            // In DEBUG mode, we prefer to exercise the code below even in the absence of extensions
+            if (extensions.Count == 0)
+            {
+                return;
+            }
+#endif
+
+            PooledHashSet<NamedTypeSymbol>? visited = null;
+            var baseType = type;
```

Comment:
> var baseType = type; [](http://example.com/codeflow?start=11&length=21)

A couple of comments:
1. It looks like we are not handling scenario when ```type``` is a generic type parameter.
2. It is not clear to me why we are not using the same approach that ```ReducedExtensionMethodSymbol.Create``` use to check compatibility and a general type argument inference algorithm like in ```ReducedExtensionMethodSymbol.InferExtensionMethodTypeArguments```. We have a parameter and we have a type for the corresponding argument. What is the reason to use different approach?

Reply:
The PR re-uses the algorithm from the roles branch as we'd discussed. But in email I suggested that we use method type inference here instead of type unification, which is what `ReducedExtensionMethodSymbol` does. I was planning to do that in a subsequent PR, but am happy to include that change here.

Yes, this PR doesn't handle type parameter receivers (it's explicitly blocked in places, marked with PROTOTYPE). I probably can revisit that after change to type inference.